### PR TITLE
migrate 10k users to database

### PR DESCRIPTION
**Pull request by huyluong22**
This is my data 
image
Now we go to migrate 10k users to database 
image
Sever wil respone:
image
Now we wait about 1 to 2 minutes and check the database again
image
image

### CODE REVIEW

**Comment by SonNgocTruong** [comment]
Please help comment documentation your code in English, @huyluong22

File: userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java [hunk 1]
```diff
@@ -21,6 +29,51 @@ public class UserServiceImpl implements IUserService {
     public UserServiceImpl(UserRepository userRepository) {
         this.userRepository = userRepository;
     }
+    @Async  // Annotation cho bất đồng bộ
+    public CompletableFuture<Void> importUsersFromCsv(MultipartFile file) throws IOException {
+        List<User> usersToSave = new ArrayList<>();
+
+        // Đọc file CSV
+        try (CSVReader reader = new CSVReader(new InputStreamReader(file.getInputStream()))) {
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java:37]: Using with BufferedReader for better memory management 
```suggestion
        try (CSVReader reader = new CSVReader(new BufferedReader(new InputStreamReader(new FileInputStream("emps.csv"), "UTF-8")), ',');
)) {
```

File: userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java [hunk 2]
```diff
@@ -21,6 +29,51 @@ public class UserServiceImpl implements IUserService {
...
+                    usersToSave.add(user);
+                    // Khi danh sách đủ 10.000 người dùng thì lưu
+                    if (usersToSave.size() == 5000) {
+                        userRepository.saveAll(usersToSave);
+                        usersToSave.clear(); // Xóa danh sách sau khi lưu
+                    }
+                } catch (ArrayIndexOutOfBoundsException | IllegalArgumentException e) {
+                    // Bỏ qua dòng có lỗi hoặc không đầy đủ
+                    System.out.println("Invalid data in CSV line, skipping this row.");
+                }
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java:65]: When you try for particular exceptions, you also need to catch the general `Exception`.
```suggestion
                } catch (ArrayIndexOutOfBoundsException | IllegalArgumentException e) {
                    // Bỏ qua dòng có lỗi hoặc không đầy đủ
                    System.out.println("Invalid data in CSV line, skipping this row. Error: " + e.getMessage());
                }catch (Exception ex) {
                    // Bỏ qua dòng có lỗi hoặc không đầy đủ
                    System.out.println("Invalid data in CSV line, skipping this row. Error: " + e.getMessage());
                }
```

File: userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java [hunk 3]
```diff
@@ -21,6 +29,51 @@ public class UserServiceImpl implements IUserService {
     public UserServiceImpl(UserRepository userRepository) {
         this.userRepository = userRepository;
     }
+    @Async  // Annotation cho bất đồng bộ
+    public CompletableFuture<Void> importUsersFromCsv(MultipartFile file) throws IOException {
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java:33]: Again, please change the method name to be readable
```suggestion
    public CompletableFuture<Void> importUsersFromCsvAsync(MultipartFile file) throws IOException {
```

**Comment by SonNgocTruong** [comment]
Also, please check the logic. What if you upload the same file (with the same content), will the data be duplicated?
If so, you need to prevent this issue.

**Comment by huyluong22** [comment]
Thank you for your question, it helped me realize another issue that needed to be addressed. With the old code, when we uploaded the file again, it would cause data duplication, so I have worked on resolving that drawback.
These are the user details located at the end of my data
image
Now I will add another user to the CSV file with the ID 10002
image
Then Choose file in swagger and execute
image
This is log in IDE
image
Now we check the database again
image

File: userservice/src/main/java/com/infodation/userservice/repositories/UserRepository.java [hunk 4]
```diff
@@ -17,9 +17,10 @@ public interface UserRepository extends JpaRepository<User, Long> {
             "WHERE :name = '' OR u.firstName LIKE %:name% OR u.lastName LIKE %:name%")
     Page<User> findByName(@Param("name") String name, Pageable pageable);
 
+    @Query("SELECT u.userId FROM users u")
+    List<String> findAllUserIds();
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/repositories/UserRepository.java:21]: You can return the Set<String> directly here, so then you no need to cast again `List` to `Set` at https://github.com/thaibinh156/user-service/pull/11/files#diff-d6ab78c7eb4c57c14c360014ef30971bc491c852e8a7caa8557c30793411363cR33-R34

File: userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java [hunk 5]
```diff
@@ -21,7 +26,65 @@ public class UserServiceImpl implements IUserService {
...
+                    user.setFirstName(nextLine[2]); // Column 2 is firstName
+                    user.setLastName(nextLine[3]); // Column 3 is lastName
+                    user.setEmail(nextLine[5]); // Column 5 is email
+                    user.setSex(Sex.valueOf(nextLine[4].toUpperCase())); // Column 4 is MALE/FEMALE
+                    user.setCreatedAt(LocalDateTime.now());
+                    user.setUpdatedAt(LocalDateTime.now());
+                    // Add user to the list if it doesn't exist in the database
+                    usersToSave.add(user);
+                    // Save when the list reaches 5,000 users
+                    if (usersToSave.size() == 5000) {
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java:67]: It is good to define this as a constant variable or define as a configuration of the application.
But I prefer setting this as a configuration (in `application.properties` file) and you can you annotation `@Value` of Springboot framework to get in your Service class.
Example: https://www.baeldung.com/spring-value-annotation

**Review by SonNgocTruong — APPROVED** [review]
LGTM

**Pull request merged**